### PR TITLE
feat(engine-core): add ENABLE_FROZEN_TEMPLATE, deep-freeze/warn

### PR DESCRIPTION
**Pull request by nolanlawson**
## Details

Partially addresses #2826.

Adds a flag that freezes the template object. This flag cannot be enabled today in core because of backwards compatibility. But potentially in the future, or in other environments, we can turn the flag on to disable this legacy behavior.

## Does this pull request introduce a breaking change?

* ✅ No, it does not introduce a breaking change.

## Does this pull request introduce an observable change?

* ✅ No, it does not introduce an observable change.

## Gus WI

[W-11111178](https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07EE00000wXyahYAC/view)

### CODE REVIEW

File: packages/@lwc/engine-core/src/framework/freeze-template.ts [hunk 1]
```diff
@@ -81,48 +83,85 @@ function warnOnArrayMutation(stylesheets: TemplateStylesheetFactories) {
...
+                        `Dynamically setting the "stylesheetTokens" property on a template function ` +
+                            `is deprecated and may be removed in a future version of LWC.`
+                    );
+                    // Avoid logging twice (for both stylesheetToken and stylesheetTokens)
+                    mutationWarningsSilenced = true;
+                    originalDescriptor!.set!.call(this, value);
+                    mutationWarningsSilenced = false;
+                },
+            });
+        }
```
Comment (nolanlawson) [inline packages/@lwc/engine-core/src/framework/freeze-template.ts:165]: This diff is really confusing. I didn't actually change this code; I just moved code from `secure-template.ts` to this file.

**Comment by nolanlawson** [comment]
/nucleus ignore --reason "lwc-platform is borked for now"

File: packages/@lwc/engine-core/src/framework/freeze-template.ts [hunk 2]
```diff
@@ -81,48 +83,85 @@ function warnOnArrayMutation(stylesheets: TemplateStylesheetFactories) {
...
-                        );
-                    }
-                    value = newValue;
-                },
-            });
+    if (features.ENABLE_FROZEN_TEMPLATE) {
+        // Deep freeze the template
+        freeze(tmpl);
+        if ('stylesheets' in tmpl) {
+            freeze(tmpl.stylesheets);
```
Comment (nolanlawson) [inline packages/@lwc/engine-core/src/framework/freeze-template.ts:91]: PR feedback: this needs to be deep-frozen (array of arrays).

**Comment by nolanlawson** [comment]
@rui-rayqiu @jmsjtu Based on our conversation today, I've made these changes to this PR:

1. Deep-freeze (and deep-warn!) on `tmpl.stylesheets` arrays
2. Freeze and warn on mutations of the `stylesheet` function itself (e.g. `stylesheet.$scoped$ = true`)

#<span></span>2 is something else I missed when I first implemented `freezeTemplate`. You could imagine someone trying to make a scoped stylesheet non-scoped, or vice versa. Today this is possible. 🤦‍♂️

**Review by jmsjtu — APPROVED** [review]
lgtm!

**Pull request merged**